### PR TITLE
Added support for connecting to TLS enabled memcached (new PR)

### PR DESCRIPTION
**Pull request by paralainer**
Picking up https://github.com/spotify/folsom/pull/202 as it seems to be abandoned, I don't have permission to push into that branch so created a new PR for that.

At this stage I merged master into the branch. I saw some intermittent test failures were reported I'll look into that as well

### CODE REVIEW

**Comment by paralainer** [comment]
PR is ready for review. Tests are ok locally but I need approval to run them on CI

File: folsom/src/test/java/com/spotify/folsom/IntegrationTest.java
Comment (paralainer) [inline folsom/src/test/java/com/spotify/folsom/IntegrationTest.java:1]: Compared the diff with AbstractIntegrationTestBase.java and confirmed that all tests are up to date

**Comment by paralainer** [comment]
@spkrka @protocol7 You seem to be on the previous PR for TLS support, can you please have a look at this one?
We really need TLS support and team is already considering vendoring of the library, which we obviously would like to avoid :)

File: folsom/src/main/java/com/spotify/folsom/client/DefaultRawMemcacheClient.java
Comment (protocol7) [inline folsom/src/main/java/com/spotify/folsom/client/DefaultRawMemcacheClient.java]: Do we need the extra constructor here? This is not part of the public API so would be fine to change the existing one.

Comment (paralainer) [reply folsom/src/main/java/com/spotify/folsom/client/DefaultRawMemcacheClient.java]: Makes sense, fixed

File: folsom/src/main/java/com/spotify/folsom/client/DefaultRawMemcacheClient.java
Comment (paralainer) [inline folsom/src/main/java/com/spotify/folsom/client/DefaultRawMemcacheClient.java:132]: Did a bit of refactoring for this logic comparing to previous PR, less code

File: folsom/src/main/java/com/spotify/folsom/reconnect/ReconnectingClient.java
Comment (paralainer) [inline folsom/src/main/java/com/spotify/folsom/reconnect/ReconnectingClient.java]: @protocol7 does it make sense to remove this constructor as well?

Comment (paralainer) [reply folsom/src/main/java/com/spotify/folsom/reconnect/ReconnectingClient.java]: It was unused and not part of public API, so removed it

File: folsom/src/main/java/com/spotify/folsom/client/tls/DefaultSSLEngineFactory.java
Comment (protocol7) [inline folsom/src/main/java/com/spotify/folsom/client/tls/DefaultSSLEngineFactory.java]: Make parameters final

Comment (paralainer) [reply folsom/src/main/java/com/spotify/folsom/client/tls/DefaultSSLEngineFactory.java]: done

File: folsom/src/test/java/com/spotify/folsom/MemcachedServer.java
Comment (protocol7) [inline folsom/src/test/java/com/spotify/folsom/MemcachedServer.java]:
```suggestion
        new FixedHostPortGenericContainer("bitnami/memcached:" + MEMCACHED_VERSION);
```

File: folsom/src/test/java/com/spotify/folsom/MemcachedServer.java
Comment (protocol7) [inline folsom/src/test/java/com/spotify/folsom/MemcachedServer.java:67]: Could this be NONE when not using TLS?

Comment (paralainer) [reply folsom/src/test/java/com/spotify/folsom/MemcachedServer.java:67]: It could, but there is no way to make this check exhaustive (fail in compile time) and for NONE we don't modify the state, so it didn't seem useful to have it. I'll return it though if you prefer to have it.

Comment (paralainer) [reply folsom/src/test/java/com/spotify/folsom/MemcachedServer.java:67]: @protocol7 any concerns here? Something else preventing this PR to be merged? 
Also can we run a CI build to make sure it's all good

File: folsom/src/main/java/com/spotify/folsom/client/DefaultRawMemcacheClient.java
Comment (ghost) [inline folsom/src/main/java/com/spotify/folsom/client/DefaultRawMemcacheClient.java]: Try to avoid wildcard imports

File: folsom/src/main/java/com/spotify/folsom/client/DefaultRawMemcacheClient.java
Comment (ghost) [inline folsom/src/main/java/com/spotify/folsom/client/DefaultRawMemcacheClient.java:120]: I strongly recommend overloading here, even if the original method just calls this new one with an extra `null` in the end. This is because this change without overloading would force library users to change their code to add an extra parameter.

File: folsom/src/main/java/com/spotify/folsom/client/tls/DefaultSSLEngineFactory.java
Comment (ghost) [inline folsom/src/main/java/com/spotify/folsom/client/tls/DefaultSSLEngineFactory.java:7]: Will this interface need more than this one implementation?

File: folsom/src/test/java/com/spotify/folsom/MemcachedServer.java
Comment (ghost) [inline folsom/src/test/java/com/spotify/folsom/MemcachedServer.java:78]: Is this actually true? I can't test it right away, but taking a superficial look at the memcached code, TLS does not disable authentication.

Comment (paralainer) [reply folsom/src/test/java/com/spotify/folsom/MemcachedServer.java:78]: That is only about that test class, the issue is with Docker image. "memcached" image supports TLS but not authentication, "bitnami/memcached" supports authentication but not TLS, I was not able to find anything remotely popular that supports both.
I tried to create a Dockerfile based on "memcached" image but enabling SASL authentication doesn't seem trivial and steals the focus from what this PR is aiming to achive, I was going to propose another PR to improve tests after this one is merged.

File: folsom/src/main/java/com/spotify/folsom/client/tls/DefaultSSLEngineFactory.java
Comment (paralainer) [reply folsom/src/main/java/com/spotify/folsom/client/tls/DefaultSSLEngineFactory.java:7]: Yes, users might want to implement their own way of constructing `SSLEngine`. The most common way is in `DefaultSSLEngineFactory` but this is not the only way.

File: folsom/src/main/java/com/spotify/folsom/client/DefaultRawMemcacheClient.java
Comment (paralainer) [reply folsom/src/main/java/com/spotify/folsom/client/DefaultRawMemcacheClient.java:120]: I was following @protocol7 's recommendations when I removed an extra constructor. The point is that constructor is not a part of public API, also it's highly unlikely one would construct `DefaultRawMemcacheClient` manually, builder gives a lot of extra features like retries and Ketama support

File: folsom-elasticache/src/main/java/com/spotify/folsom/elasticache/ElastiCacheResolver.java
Comment (paralainer) [inline folsom-elasticache/src/main/java/com/spotify/folsom/elasticache/ElastiCacheResolver.java:202]: This is a bit opionated but given that the purpose of this resolver is to connect to AWS resource alternative methods won't be required anyway.
I checked that on a real application and it works perfectly fine.

**Comment by paralainer** [comment]
@thiagocesarj thanks! Can you triger the build?

**Comment by paralainer** [comment]
@thiagocesarj I think I found the reason for test failures - SSL data aggregation doesn't play well with memcached protocol.
Basically what happens is memcached is waiting for client to send some data to finish an operation but client doesn't send data before it has enough to fill size defined in `wrapDataSize`
Can you please run the build again, it seems to work fine for my local branch/workflow.

**Comment by thiagocesarj** [comment]
@paralainer seems like we have a successful build. I will leave the PR open for a couple more days and ask a couple of people to take a second look, and if it's all good, I will prepare a release. Thanks for the contribution!

**Review by thiagocesarj — APPROVED** [review]

**Comment by thiagocesarj** [comment]
There was an issue with a test during the master build. I will investigate, but for now, the PR has been reverted.

**Comment by paralainer** [comment]
> There was an issue with a test during the master build. I will investigate, but for now, the PR has been reverted.

@thiagocesarj I'll have a look into build failure and raise another PR with a fix

**Pull request merged**
